app/server/config/loader.ts

```
@@ -1,82 +1,106 @@
+	const rawData = await loadConfigFile(path);
+	if (typeof rawData !== 'object' || rawData === null) {
+		throw new ConfigError('Loaded configuration data is not a valid object.');
 	}
 
+	// 1. Initial validation
+	const initialValidatedConfig = validateConfig({
+		...(rawData as Record<string, unknown>),
+		debug: log.debugEnabled,
+	});
+
+	// Deep clone before mutation by loadSecretsFromFiles
+	let configObject = JSON.parse(JSON.stringify(initialValidatedConfig));
+
+	// Process *_path fields from the YAML file itself, regardless of env loading
+	await loadSecretsFromFiles(configObject);
+
 	if (!loadEnv) {
 		log.debug('config', 'Environment variable overrides are disabled');
 		log.debug('config', 'This also disables the loading of a .env file');
-		return config;
+		return validateConfig(configObject); // Re-validate after potential modifications by loadSecretsFromFiles
 	}
 
 	log.info('config', 'Loading a .env file (if available)');
 	configDotenv({ override: true });
-	config = coalesceEnv(config);
-	if (!config) {
-		exit(1);
-	}
 
-	return config;
+	configObject = coalesceEnv(configObject);
+
+	await loadSecretsFromFiles(configObject);
+
+	return validateConfig(configObject);
 }
 
 export async function hp_loadConfig() {
-	// 	// OIDC Related Checks
-	// 	if (config.oidc) {
-	// 		if (!config.oidc.client_secret && !config.oidc.client_secret_path) {
-	// 			log.error('CFGX', 'OIDC configuration is missing a secret, disabling');
-	// 			log.error(
-	// 				'CFGX',
-	// 				'Please specify either `oidc.client_secret` or `oidc.client_secret_path`',
-	// 			);
-	// 		}
-	// 		if (config.oidc?.strict_validation) {
-	// 			const result = await testOidc(config.oidc);
-	// 			if (!result) {
-	// 				log.error('CFGX', 'OIDC configuration failed validation, disabling');
-	// 			}
-	// 		}
-	// 	}
+	// This will be implemented in the future
 }
 
-async function validateConfigPath(path: string) {
+async function validateConfigPath(path: string): Promise<void> {
```

Is there a reason you chose to begin throwing a ConfigError here instead of just handling the error and returning a boolean? I don't necessarily think the access failure message is useful here.

---

Yeah that was something I was debating. It was useful for debugging for me as I got explicit error messages for certain conditions. But I could also change it to just print an error message and return a bool. However if validation fails the program should just exit anyways no?

---

Yes, we can keep it as a ConfigError as long as its formatted well on stdout.

---

I'm going with a hybrid approach. I agree it should return a bool and let the caller deal with it as needed.